object.c

```diff
@@ -4681,6 +4686,256 @@ InitVM_Object(void)
+            {
+                uint32_t len = ROBJECT_NUMIV(obj);
+                VALUE *ptr = ROBJECT_IVPTR(obj);
+
+                for (uint32_t i=0; i<len; i++) {
+                    VALUE val = ptr[i];
+                    if (val != Qundef) {
+                        if (rb_obj_traverse_i(val, func, data, rec)) {
+                            return 1;
+                        }
+                    }
+                }
+            }
+            break;
+
+          case T_ARRAY:
+            {
+                for (int i = 0; i < RARRAY_LENINT(obj); i++) {
+                    VALUE e = rb_ary_entry(obj, i);
+                    if (rb_obj_traverse_i(e, func, data, rec)) {
+                        return 1;
+                    }
+                }
+            }
+            break;
+
+          case T_HASH:
+            {
+                struct obj_hash_traverse_data d = {
+                    .stop = 0,
+                    .func = func,
+                    .data = data,
+                    .rec  = rec,
+                };
+                rb_hash_foreach(obj, obj_hash_traverse_i, (VALUE)&d);
+                if (d.stop) return 1;
+            }
+            break;
+
+          case T_STRUCT:
+            {
+                long len = RSTRUCT_LEN(obj);
+                const VALUE *ptr = RSTRUCT_CONST_PTR(obj);
+
+                for (long i=0; i<len; i++) {
+                    if (rb_obj_traverse_i(ptr[i], func, data, rec)) {
+                        return 1;
+                    }
+                }
+            }
+            break;
+
+          case T_RATIONAL:
+            if (rb_obj_traverse_i(RRATIONAL(obj)->num, func, data, rec)) return 1;
+            if (rb_obj_traverse_i(RRATIONAL(obj)->den, func, data, rec)) return 1;
+            break;
+          case T_COMPLEX:
+            if (rb_obj_traverse_i(RCOMPLEX(obj)->real, func, data, rec)) return 1;
+            if (rb_obj_traverse_i(RCOMPLEX(obj)->imag, func, data, rec)) return 1;
+            break;
+
+          case T_DATA:
+            {
+                // TODO
+            }
+            break;
+
+          case T_CLASS:
+          case T_MODULE:
+          case T_ICLASS:
+            // TODO
+            break;
+
+          default:
+            rp(obj);
+            rb_bug("unreachable");
+        }
+    }
+    return 0;
+}
+
+struct obj_traverse_iterate_data {
+    rb_obj_traverse_iterate_func func;
+    void *data;
+};
+
+static int
+obj_traverse_iterate_i(st_data_t key, st_data_t val, st_data_t ptr)
+{
+    struct obj_traverse_iterate_data *data = (struct obj_traverse_iterate_data *)ptr;
+
+    if (data->func(key, data->data)) {
+        return ST_STOP;
+    }
+    else {
+        return ST_CONTINUE;
+    }
+}
+
+// 0: traverse all
+// 1: stopped
+int
+rb_obj_traverse(VALUE obj,
+                rb_obj_traverse_func tfunc, void *tdata,
+                rb_obj_traverse_iterate_func iter_func, void *iter_data)
+{
+    st_table *rec = st_init_numtable();
```

Comment:
Can we do this instead?

```suggestion
    VALUE wrapper = rb_ident_hash_new();
    st_table *rec = rb_hash_st_table(wrapper);
```

Then the GC will not move references stored in `rec`.  Also I think we can delete the calls to `st_free_table` below.

Reply:
awesome.